### ISSUE TITLE
Build breaks with OPC UA SDK 1.4.3

### ISSUE THREAD

**Issue by robi-wan** [uaf/uaf#51]
Hi Wim,

I tried to build uaf (on Windows with VS 2010) with the current version of OPC UA SDK 1.4.3 - with errors:
```
[CL] helperfunctions.cpp
[CL] C:\Programme\Microsoft SDKs\Windows\v7.0A\include\ws2def.h(91, 0): warning C4005: 'AF_IPX' : macro redefinition
[CL]         C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock.h(460) : see previous definition of 'AF_IPX'
[CL] C:\Programme\Microsoft SDKs\Windows\v7.0A\include\ws2def.h(131, 0): warning C4005: 'AF_MAX' : macro redefinition
[CL]         C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock.h(479) : see previous definition of 'AF_MAX'
[CL] C:\Programme\Microsoft SDKs\Windows\v7.0A\include\ws2def.h(168, 0): warning C4005: 'SO_DONTLINGER' : macro redefinition
[CL]         C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock.h(402) : see previous definition of 'SO_DONTLINGER'
[CL] C:\Programme\Microsoft SDKs\Windows\v7.0A\include\ws2def.h(212, 0): error C2011: 'sockaddr' : 'struct' type redefinition
[CL]         C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock.h(485) : see declaration of 'sockaddr'
[CL] C:\Programme\Microsoft SDKs\Windows\v7.0A\include\ws2def.h(390, 0): error C2059: syntax error : 'constant'
[CL] C:\Programme\Microsoft SDKs\Windows\v7.0A\include\ws2def.h(390, 0): error C3805: 'constant': unexpected token, expected either '}' or a ','
[CL] C:\Programme\Microsoft SDKs\Windows\v7.0A\include\ws2def.h(524, 0): warning C4005: 'IN_CLASSA' : macro redefinition
[CL]         C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock.h(287) : see previous definition of 'IN_CLASSA'
[CL] C:\Programme\Microsoft SDKs\Windows\v7.0A\include\ws2def.h(530, 0): warning C4005: 'IN_CLASSB' : macro redefinition
[CL]         C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock.h(293) : see previous definition of 'IN_CLASSB'
[CL] C:\Programme\Microsoft SDKs\Windows\v7.0A\include\ws2def.h(536, 0): warning C4005: 'IN_CLASSC' : macro redefinition
[CL]         C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock.h(299) : see previous definition of 'IN_CLASSC'
[CL] C:\Programme\Microsoft SDKs\Windows\v7.0A\include\ws2def.h(547, 0): warning C4005: 'INADDR_ANY' : macro redefinition
[CL]         C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock.h(304) : see previous definition of 'INADDR_ANY'
[CL] C:\Programme\Microsoft SDKs\Windows\v7.0A\include\ws2def.h(549, 0): warning C4005: 'INADDR_BROADCAST' : macro redefinition
[CL]         C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock.h(306) : see previous definition of 'INADDR_BROADCAST'
[CL] C:\Programme\Microsoft SDKs\Windows\v7.0A\include\ws2def.h(583, 0): error C2011: 'sockaddr_in' : 'struct' type redefinition
[CL]         C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock.h(312) : see declaration of 'sockaddr_in'
[CL] C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock2.h(132, 0): error C2011: 'fd_set' : 'struct' type redefinition
[CL]         C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock.h(68) : see declaration of 'fd_set'
[CL] C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock2.h(167, 0): warning C4005: 'FD_SET' : macro redefinition
[CL]         C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock.h(102) : see previous definition of 'FD_SET'
[CL] C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock2.h(176, 0): error C2011: 'timeval' : 'struct' type redefinition
[CL]         C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock.h(111) : see declaration of 'timeval'
[CL] C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock2.h(232, 0): error C2011: 'hostent' : 'struct' type redefinition
[CL]         C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock.h(167) : see declaration of 'hostent'
[CL] C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock2.h(245, 0): error C2011: 'netent' : 'struct' type redefinition
[CL]         C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock.h(180) : see declaration of 'netent'
[CL] C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock2.h(252, 0): error C2011: 'servent' : 'struct' type redefinition
[CL]         C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock.h(187) : see declaration of 'servent'
[CL] C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock2.h(264, 0): error C2011: 'protoent' : 'struct' type redefinition
[CL]         C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock.h(199) : see declaration of 'protoent'
[CL] C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock2.h(360, 0): error C2011: 'WSAData' : 'struct' type redefinition
[CL]         C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock.h(322) : see declaration of 'WSAData'
[CL] C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock2.h(458, 0): error C2011: 'sockproto' : 'struct' type redefinition
[CL]         C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock.h(494) : see declaration of 'sockproto'
[CL] C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock2.h(500, 0): error C2011: 'linger' : 'struct' type redefinition
[CL]         C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock.h(531) : see declaration of 'linger'
[CL] C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock2.h(513, 0): warning C4005: 'SOMAXCONN' : macro redefinition
[CL]         C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock.h(544) : see previous definition of 'SOMAXCONN'
[CL] C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock2.h(542, 0): warning C4005: 'FD_READ' : macro redefinition
[CL]         C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock.h(562) : see previous definition of 'FD_READ'
[CL] C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock2.h(545, 0): warning C4005: 'FD_WRITE' : macro redefinition
[CL]         C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock.h(563) : see previous definition of 'FD_WRITE'
[CL] C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock2.h(548, 0): warning C4005: 'FD_OOB' : macro redefinition
[CL]         C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock.h(564) : see previous definition of 'FD_OOB'
[CL] C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock2.h(551, 0): warning C4005: 'FD_ACCEPT' : macro redefinition
[CL]         C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock.h(565) : see previous definition of 'FD_ACCEPT'
[CL] C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock2.h(554, 0): warning C4005: 'FD_CONNECT' : macro redefinition
[CL]         C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock.h(566) : see previous definition of 'FD_CONNECT'
[CL] C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock2.h(557, 0): warning C4005: 'FD_CLOSE' : macro redefinition
[CL]         C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock.h(567) : see previous definition of 'FD_CLOSE'
[CL] C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock2.h(1578, 0): error C2375: 'accept' : redefinition; different linkage
[CL]         C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock.h(742) : see declaration of 'accept'
[CL] C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock2.h(1600, 0): error C2375: 'bind' : redefinition; different linkage
[CL]         C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock.h(747) : see declaration of 'bind'
[CL] C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock2.h(1619, 0): error C2375: 'closesocket' : redefinition; different linkage
[CL]         C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock.h(752) : see declaration of 'closesocket'
[CL] C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock2.h(1638, 0): error C2375: 'connect' : redefinition; different linkage
[CL]         C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock.h(754) : see declaration of 'connect'
[CL] C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock2.h(1659, 0): error C2375: 'ioctlsocket' : redefinition; different linkage
[CL]         C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock.h(759) : see declaration of 'ioctlsocket'
[CL] C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock2.h(1680, 0): error C2375: 'getpeername' : redefinition; different linkage
[CL]         C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock.h(764) : see declaration of 'getpeername'
[CL] C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock2.h(1701, 0): error C2375: 'getsockname' : redefinition; different linkage
[CL]         C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock.h(769) : see declaration of 'getsockname'
[CL] C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock2.h(1724, 0): error C2375: 'getsockopt' : redefinition; different linkage
[CL]         C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock.h(774) : see declaration of 'getsockopt'
[CL] C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock2.h(1745, 0): error C2375: 'htonl' : redefinition; different linkage
[CL]         C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock.h(781) : see declaration of 'htonl'
[CL] C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock2.h(1762, 0): error C2375: 'htons' : redefinition; different linkage
[CL]         C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock.h(783) : see declaration of 'htons'
[CL] C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock2.h(1780, 0): error C2375: 'inet_addr' : redefinition; different linkage
[CL]         C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock.h(785) : see declaration of 'inet_addr'
[CL] C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock2.h(1797, 0): error C2375: 'inet_ntoa' : redefinition; different linkage
[CL]         C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock.h(787) : see declaration of 'inet_ntoa'
[CL] C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock2.h(1815, 0): error C2375: 'listen' : redefinition; different linkage
[CL]         C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock.h(789) : see declaration of 'listen'
[CL] C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock2.h(1833, 0): error C2375: 'ntohl' : redefinition; different linkage
[CL]         C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock.h(793) : see declaration of 'ntohl'
[CL] C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock2.h(1850, 0): error C2375: 'ntohs' : redefinition; different linkage
[CL]         C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock.h(795) : see declaration of 'ntohs'
[CL] C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock2.h(1870, 0): error C2375: 'recv' : redefinition; different linkage
[CL]         C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock.h(797) : see declaration of 'recv'
[CL] C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock2.h(1895, 0): error C2375: 'recvfrom' : redefinition; different linkage
[CL]         C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock.h(803) : see declaration of 'recvfrom'
[CL] C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock2.h(1921, 0): error C2375: 'select' : redefinition; different linkage
[CL]         C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock.h(811) : see declaration of 'select'
[CL] C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock2.h(1945, 0): error C2375: 'send' : redefinition; different linkage
[CL]         C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock.h(818) : see declaration of 'send'
[CL] C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock2.h(1970, 0): error C2375: 'sendto' : redefinition; different linkage
[CL]         C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock.h(824) : see declaration of 'sendto'
[CL] C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock2.h(1996, 0): error C2375: 'setsockopt' : redefinition; different linkage
[CL]         C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock.h(832) : see declaration of 'setsockopt'
[CL] C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock2.h(2018, 0): error C2375: 'shutdown' : redefinition; different linkage
[CL]         C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock.h(839) : see declaration of 'shutdown'
[CL] C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock2.h(2039, 0): error C2375: 'socket' : redefinition; different linkage
[CL]         C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock.h(843) : see declaration of 'socket'
[CL] C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock2.h(2063, 0): error C2375: 'gethostbyaddr' : redefinition; different linkage
[CL]         C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock.h(850) : see declaration of 'gethostbyaddr'
[CL] C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock2.h(2083, 0): error C2375: 'gethostbyname' : redefinition; different linkage
[CL]         C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock.h(855) : see declaration of 'gethostbyname'
[CL] C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock2.h(2101, 0): error C2375: 'gethostname' : redefinition; different linkage
[CL]         C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock.h(857) : see declaration of 'gethostname'
[CL] C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock2.h(2121, 0): error C2375: 'getservbyport' : redefinition; different linkage
[CL]         C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock.h(861) : see declaration of 'getservbyport'
[CL] C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock2.h(2141, 0): error C2375: 'getservbyname' : redefinition; different linkage
[CL]         C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock.h(865) : see declaration of 'getservbyname'
[CL] C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock2.h(2160, 0): error C2375: 'getprotobynumber' : redefinition; different linkage
[CL]         C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock.h(869) : see declaration of 'getprotobynumber'
[CL] C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock2.h(2178, 0): error C2375: 'getprotobyname' : redefinition; different linkage
[CL]         C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock.h(871) : see declaration of 'getprotobyname'
[CL] C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock2.h(2200, 0): error C2375: 'WSAStartup' : redefinition; different linkage
[CL]         C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock.h(875) : see declaration of 'WSAStartup'
[CL] C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock2.h(2219, 0): error C2375: 'WSACleanup' : redefinition; different linkage
[CL]         C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock.h(879) : see declaration of 'WSACleanup'
[CL] C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock2.h(2236, 0): error C2375: 'WSASetLastError' : redefinition; different linkage
[CL]         C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock.h(881) : see declaration of 'WSASetLastError'
[CL] C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock2.h(2253, 0): error C2375: 'WSAGetLastError' : redefinition; different linkage
[CL]         C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock.h(883) : see declaration of 'WSAGetLastError'
[CL] C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock2.h(2270, 0): error C2375: 'WSAIsBlocking' : redefinition; different linkage
[CL]         C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock.h(885) : see declaration of 'WSAIsBlocking'
[CL] C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock2.h(2287, 0): error C2375: 'WSAUnhookBlockingHook' : redefinition; different linkage
[CL]         C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock.h(887) : see declaration of 'WSAUnhookBlockingHook'
[CL] C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock2.h(2304, 0): error C2375: 'WSASetBlockingHook' : redefinition; different linkage
[CL]         C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock.h(889) : see declaration of 'WSASetBlockingHook'
[CL] C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock2.h(2321, 0): error C2375: 'WSACancelBlockingCall' : redefinition; different linkage
[CL]         C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock.h(891) : see declaration of 'WSACancelBlockingCall'
[CL] C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock2.h(2343, 0): error C2375: 'WSAAsyncGetServByName' : redefinition; different linkage
[CL]         C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock.h(893) : see declaration of 'WSAAsyncGetServByName'
[CL] C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock2.h(2370, 0): error C2375: 'WSAAsyncGetServByPort' : redefinition; different linkage
[CL]         C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock.h(901) : see declaration of 'WSAAsyncGetServByPort'
[CL] C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock2.h(2396, 0): error C2375: 'WSAAsyncGetProtoByName' : redefinition; different linkage
[CL]         C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock.h(909) : see declaration of 'WSAAsyncGetProtoByName'
[CL] C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock2.h(2421, 0): error C2375: 'WSAAsyncGetProtoByNumber' : redefinition; different linkage
[CL]         C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock.h(916) : see declaration of 'WSAAsyncGetProtoByNumber'
[CL] C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock2.h(2446, 0): error C2375: 'WSAAsyncGetHostByName' : redefinition; different linkage
[CL]         C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock.h(923) : see declaration of 'WSAAsyncGetHostByName'
[CL] C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock2.h(2473, 0): error C2375: 'WSAAsyncGetHostByAddr' : redefinition; different linkage
[CL]         C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock.h(930) : see declaration of 'WSAAsyncGetHostByAddr'
[CL] C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock2.h(2496, 0): error C2375: 'WSACancelAsyncRequest' : redefinition; different linkage
[CL]         C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock.h(939) : see declaration of 'WSACancelAsyncRequest'
[CL] C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock2.h(2517, 0): error C2375: 'WSAAsyncSelect' : redefinition; different linkage
[CL]         C:\Programme\Microsoft SDKs\Windows\v7.0A\include\winsock.h(941) : see declaration of 'WSAAsyncSelect'

```

There seems to be an issue with redefinitions: "This problem is caused when including <windows.h> before <winsock2.h>.".

More info:
http://stackoverflow.com/questions/1372480/c-redefinition-header-files-winsock2-h
http://stackoverflow.com/questions/11495589/winsock-redefinition-errors
http://stackoverflow.com/questions/4061562/vs2008-compile-error-with-winsock2-h
http://www.zachburlingame.com/2011/05/resolving-redefinition-errors-betwen-ws2def-h-and-winsock-h/

(Links include multiple solutions for this problem.)

For uaf the cause (windows.h include) is in [helperfunctions.h](https://github.com/uaf/uaf/blob/master/src/uaf/util/helperfunctions.h#L29) which includes uaf/util/util.h which includes uabase/uaplatformlayer.h which includes uabase/uaplatformdefs.h - and this last file changed between OPC UA 1.4.2 and 1.4.3:

```cpp
--- C:/development/tools/opc_ua_sdk_1.4.2/include/uabase/uaplatformdefs.h	Di Apr 29 10:56:00 2014
+++ C:/development/tools/opc_ua_sdk_1.4.3/include/uabase/uaplatformdefs.h	Di Jan 20 14:07:24 2015
@@ -1,7 +1,7 @@
 /******************************************************************************
 ** uaplatformdefs.h
 **
-** Copyright (c) 2006-2014 Unified Automation GmbH All rights reserved.
+** Copyright (c) 2006-2015 Unified Automation GmbH All rights reserved.
 **
 ** Software License Agreement ("SLA") Version 2.3
 **
@@ -40,6 +40,12 @@
 #include <malloc.h>
 #include <stdlib.h>
 #include <memory.h>
+#ifndef WIN32_LEAN_AND_MEAN
+/* Always define WIN32_LEAN_AND_MEAN to prevent windows.h from including
+ * winsock.h which leads to conflicts with winsock2.h */
+# define WIN32_LEAN_AND_MEAN
+#endif
+#include <winsock2.h>
 #include <windows.h>
 #include <oaidl.h>  // windows VARIANT definition
 #include <olectl.h> // windows FILETIME definition
@@ -79,8 +85,6 @@
     #define UABASE_EXPORT
 #endif
 
-#define snprintf _snprintf
-
 #include <tchar.h>
 #define UA_swprintf _snwprintf
 
@@ -132,9 +136,15 @@ OpcUa_Void UA_SetLastError(OpcUa_UInt32 nError);
 #define OpcUa_PRIi64 "I64d"
 #define OpcUa_PRIu64 "I64u"
 
-#define swprintf    _snwprintf
-#define snprintf    _snprintf
-#define wcsncasecmp _wcsnicmp
+#ifndef swprintf
+    #define swprintf    _snwprintf
+#endif
+#ifndef snprintf
+    #define snprintf    _snprintf
+#endif
+#ifndef wcsncasecmp
+    #define wcsncasecmp _wcsnicmp
+#endif
 
 OpcUa_Int32 OpcUa_MsecTo(OpcUa_DateTime fromTime, OpcUa_DateTime toTime);
```

I think the relevant lines are these:
```cpp
@@ -40,6 +40,12 @@
 #include <malloc.h>
 #include <stdlib.h>
 #include <memory.h>
+#ifndef WIN32_LEAN_AND_MEAN
+/* Always define WIN32_LEAN_AND_MEAN to prevent windows.h from including
+ * winsock.h which leads to conflicts with winsock2.h */
+# define WIN32_LEAN_AND_MEAN
+#endif
+#include <winsock2.h>
 #include <windows.h>
 #include <oaidl.h>  // windows VARIANT definition
 #include <olectl.h> // windows FILETIME definition
```

I tried several of the suggested solutions (http://stackoverflow.com/a/8294669) with the file helperfunctions.h:
1. add `#define _WINSOCKAPI_` before the line with #include <windows.h>
2. 
```cpp
#define WIN32_LEAN_AND_MEAN
#include <windows.h>
#undef WIN32_LEAN_AND_MEAN
```
3. replace `#include <windows.h>` with `#include <winsock2.h>`

All solutions compiled.
Another simple solution also worked: removing the line which contains `#include <windows.h>`  - as this is the only line for the if-branch of the ifdef the whole ifdef can be changed:

```cpp
 src/uaf/util/helperfunctions.h | 4 +---
 1 file changed, 1 insertion(+), 3 deletions(-)

diff --git a/src/uaf/util/helperfunctions.h b/src/uaf/util/helperfunctions.h
index 455156d..1bd58de 100644
--- a/src/uaf/util/helperfunctions.h
+++ b/src/uaf/util/helperfunctions.h
@@ -25,9 +25,7 @@
 #include <string>
 #include <algorithm>
 
-#ifdef _WIN32
-#include <windows.h>
-#else
+#ifndef _WIN32
 #include <limits.h>
 #include <unistd.h>
 #endif
```

At least it compiles.
Just tested under windows; I did not run the unit tests.

**Comment by WimPessemier** [comment]
Hi,

Ah sorry I did not find the time to test the new SDK, thanks a lot for investigating and sharing solutions!
I'll look at it over the weekend, but it seems like you already did all the hard work :-)

Thanks,
Wim

**Comment by robi-wan** [comment]
I forgot to mention: I compiled uaf with these fixes just against SDK 1.4.3 - I don't know if it also works (compiles) with older SDK 1.4.* versions.

**Comment by WimPessemier** [comment]
Hi,

I've applied your last solution (i.e. remove ``#include <windows.h>``) and tested it on Linux and Windows with SDK v1.4.2 and v1.4.3. Everything appears to be working so I've updated the master and long-term branch to include the bugfix.

Thanks again!

Wim
